DM-20842: Change Formatter API

timj opened the pull request:
The motivation for this is to simplify formatters so that they know they can store state in their instance without fear that a Datastore is going to reuse the Formatter for some other location.
    
* This changes the read() and write() methods since they no longer need FileDescriptor parameter.
* All the formatters need to be tweaked.
* Datastore needs changing to ensure that FileDescriptor is created a bit earlier.
* getFormatter gets an extra argument which can be None for formatters that will not be using read/write at all.

@TallJimbo , @parejkoj does this look okay to you?

timj committed:
Allow getInstanceOf to be given parameters for object instantiation

timj committed:
Allow MappingFactory to accept parameters for object instantiation

timj committed:
Change Formatter to require a FileDescriptor contructor argument

The motivation for this is to simplify formatters so
that they know they can store state in their instance
without fear that a Datastore is going to reuse the
Formatter for some other location.

* This changes the read() and write() methods since they no
  longer need FileDescriptor parameter.
* All the formatters need to be tweaked.
* Datastore needs changing to ensure that FileDescriptor is
  created a bit earlier.
* getFormatter gets an extra argument which can be None for
  formatters that will not be using read/write at all.

parejkoj commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -27,13 +27,20 @@
+
+    Parameters
+    ----------
+    fileDescriptor : `FileDescriptor`
```
I'm not sure, but I think you still have to give the full namespace in order for this to become a link.

timj replied:
I was under the impression that classes in the local package didn't need to be fully qualified.

parejkoj commented on tests/test_formatter.py:
```diff
@@ -45,15 +45,15 @@ def testRegistry(self):
         storageClassName = "Image"
         self.factory.registerFormatter(storageClassName, formatterTypeName)
-        f = self.factory.getFormatter(storageClassName)
+        f = self.factory.getFormatter(storageClassName, None)
```
Are all of these `None`s really necessary here? Could you make wherever that is landing take a `None` default arg?

timj replied:
I decided to force FileDescriptor to take an argument.  This at least made it obvious where I needed to put an argument during testing. It's more explicit this way that you are saying you know that the formatter will never be used to read or write anything.

parejkoj reviewed (commented):
Just two comments, otherwise this looks fine.

TallJimbo commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -42,21 +49,27 @@ class Formatter(metaclass=ABCMeta):
+    def __init__(self, fileDescriptor):
+        if fileDescriptor is not None and not isinstance(fileDescriptor, FileDescriptor):
+            raise TypeError("File descriptor must be a FileDescriptor")
+        self._fileDescriptor = fileDescriptor
```
Seems like this might as well be a public attribute.

timj replied:
No. It's a read-only property. There's a getter further down.

timj replied:
Meaning I want to guarantee that people can't change the file location once the object is constructed.

TallJimbo reviewed (approved):
:+1: I'm looking forward to seeing how this improves the concrete raw formatters; I think they'll really benefit from this change.

timj committed:
Allow Formatter constructor to be called with no arguments

timj committed:
Datastore should raise from the previous exception

timj committed:
More fixes for new Formatter usage

timj committed:
Reimplement factory to allow a class to be returned

Instance retrieval is now layered on top of class retrieval.

timj committed:
Add getFormatterClass to FormatterFactory

timj committed:
Allow storefFileInfo to take a Formatter class as well as instance

timj committed:
Add Formatter.predictPathWithLocation method

This can be a class method and so be distinct from the instance
methoer predictPath

timj committed:
Change ingest to accept a formatter class

timj committed:
Switch formatter validation to use getFormatterClass

timj commented:
Related pull requests in lsst/obs_subaru#217 and lsst/obs_base#165

I think I'm ready for a proper review. I rejigged things in the way suggested by @TallJimbo to allow for a `Formatter.getFormatterClass` method which makes the logic in datastore.ingest more explicit.

TallJimbo commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -87,17 +99,38 @@ def write(self, inMemoryDataset, fileDescriptor):
+        ----------
         location : `Location`
-            The location to simulate writing to.
+            Location of file for which path prediction is required.
```
I gather this is a root filename, and what's returned is the full thing with the extension (or some generalization thereof)?  This is mostly preexisting, but it's not currently clear from the docs what the distinction between input and output is.

timj replied:
This method doesn't care about input or output. It just wants a `Location` object which refers to a place in the datastore -- this method generally applies the right file suffix and then returns the filename from the Location.

TallJimbo commented on python/lsst/daf/butler/core/storedFileInfo.py:
```diff
@@ -50,9 +52,13 @@ class StoredFileInfo:
+                                                  issubclass(formatter, Formatter)):
             formatter = formatter.name()
+        else:
+            raise ValueError(f"Supplied formatter '{formatter}' is not supported")
```
What does "not supported" mean in this context?  That it's not actually a `Formatter`?

TallJimbo commented on python/lsst/daf/butler/formatters/fitsExposureFormatter.py:
```diff
@@ -31,8 +31,18 @@ class FitsExposureFormatter(Formatter):
+    def metadata(self):
+        """The metadata read from this file. It will be stripped as
+        components are extracted from it.
+        """
```
Docstring should include the return type.

parejkoj replied:
Almost: https://developer.lsst.io/python/numpydoc.html#documenting-class-properties

parejkoj commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -197,7 +234,57 @@ def getLookupKeys(self):
 
-    def getFormatterWithMatch(self, entity):
+    def getFormatterClassWithMatch(self, entity):
+        """Get a the matching formatter class along with the matching registry
```
get a the?

parejkoj commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -197,7 +234,57 @@ def getLookupKeys(self):
+
+        Parameters
+        ----------
+        entity : `DatasetRef`, `DatasetType` or `StorageClass`, or `str`
```
Either missing an `or`, or has one to many. Same in `getFormatterClass` below.

parejkoj reviewed (approved):
Couple of docstring comments, otherwise looks good.

timj committed:
Fix some docstrings
